src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,178 @@
+        }
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index when set node");
+        }
+        Node<T> nodeToSetNewValue = getNodeByIndex(index);
+        T oldValue = nodeToSetNewValue.value;
+        nodeToSetNewValue.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index when remove");
+        }
+        Node<T> nodeToRemove = getNodeByIndex(index);
+        unlink(index);
+        size--;
+        return nodeToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        int index = 0;
+        while (node != null) {
+            if (node.value == object || node.value != null && node.value.equals(object)) {
+                unlink(index);
+                size--;
+                return true;
+            }
+            index++;
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> nodeByIndex;
+        if (index < size >> 1) {
+            nodeByIndex = head;
+            for (int i = 0; i < index; i++) {
+                nodeByIndex = nodeByIndex.next;
+            }
+        } else {
+            nodeByIndex = tail;
+            for (int i = size - 1; i > index; i--) {
+                nodeByIndex = nodeByIndex.prev;
+            }
+        }
+        return nodeByIndex;
+    }
+
+    private void linkFirst(T value) {
+        Node<T> newNode = new Node<>(value, null, null);
+        if (head == null) {
+            head = newNode;
+            tail = head;
+        } else {
+            head.prev = newNode;
+            head = newNode;
+        }
+    }
+
+    private void linkLast(T value) {
+        Node<T> lastNode = tail;
+        Node<T> newNode = new Node<>(value, lastNode, null);
+        tail = newNode;
+        if (lastNode == null) {
+            head = newNode;
+        } else {
+            lastNode.next = newNode;
+        }
+    }
+
+    private void linkBefore(T value, Node<T> indexNode) {
+        Node<T> currentIndexNode = indexNode.prev;
+        Node<T> currentNewNode = new Node<>(value, currentIndexNode, indexNode);
+        indexNode.prev = currentNewNode;
+        if (currentIndexNode == null) {
+            head = currentNewNode;
+        } else {
+            currentIndexNode.next = currentNewNode;
+        }
+    }
+
+    private void unlink(int index) {
+        Node<T> currentNode = getNodeByIndex(index);
```

Comment:
```suggestion
    private void unlink(Node<T> currentNode) {
```
let's pass the Node to be removed as a method parameter
then we can fix next comment